### ISSUE TITLE
Low 32-bits of seed data is discarded for h2 in MurmurHash64B

### ISSUE THREAD

**Issue by bryc** [aappleby/smhasher#56]
The [MurmurHash64B function of MurmurHash2](https://github.com/aappleby/smhasher/blob/master/src/MurmurHash2.cpp#L142) specifies a 64-bit seed like so: 

```js
uint64_t MurmurHash64B ( const void * key, int len, uint64_t seed )
```

Then proceeds to shift the high 32-bits over to initialize  `h2`:

```js
uint32_t h2 = uint32_t(seed >> 32);
```

This means that no matter what the seed is, if `seed < 0x100000000`, then `h2` will always be initialized to 0. Effectively making h1 and h2 use two different 32-bit seed components split from a 64-bit integer. 

**Is there a sound reason that `h2` cannot use the same seed as `h1` (even a transformation)?**

To implement this function without support for 64-bit integer shifts, the arguments must be equivalent to `uint32_t seed1` and `uint32_t seed2` instead of just one `uint32_t seed`. 

Otherwise `(seed >> 32) === seed` and the hash value will not match, making it not fully 32-bit compatible (using JavaScript port which uses 32-bit shifts).

**Comment by sebres** [comment]
> Is there a sound reason that h2 cannot use the same seed as h1 (even a transformation)?

Yes, because the distribution is a bit "smoother" for data that are longer as 64-bit (for which purposes the hash should be used). And it's so better for longer seed values.
I'm fun :smile: 
The only reason, if you'll change the algorithm it will be not MMH anymore, so not backwards-compatible and not usable on some systems or platforms with it.
Of course if you want to use it only for your own purposes, you can adjust it how you want.

I assume, you don't understand the real purposes of the usage of the seed.

- for example, you can easy hash different families of the data-sets (possibly with the same binary form), but which should produce different hashes (even using the same hash-function), e.g. dog "Charlie" and cat "Charlie" becomes different hashes, so are stored with large probability in different buckets of the hash-table.
- or if you'll find that your input data produce badly distributed hashes, you can choose a new (e. g. any random) seed value and rehash the data (e. g. via config-value, without algorithm change, recompile etc.).
- or to store the different values for different users (each user has own seed).

But if I would answer shortly - seed is just a seed, so the values < 0x100000000 are just as well as the other seeds (so 0x0 as well as 0x5555555555555555, 0xaaaaaaaaaaaaaaaa etc).
Ultimately it's important that the resulting hash changes, if the seed changes.

Nevertheless if it bothers you, just use any "full" 64-bit seed, like `0xde59dbf86f8bd67c`.

Anyway for hash-protection in any cryptographic sense, it's not advisable to use only the seed alone, as already answered in https://github.com/aappleby/smhasher/issues/57#issuecomment-380769637

**Comment by bryc** [comment]
> I assume, you don't understand the real purposes of the usage of the seed.

No, I understand what a seed is. It is a parameter that alters the initial state, so same inputs give different outputs.

There is simply confusion because `MurmurHash64B` is a 32-bit hash function, but its seed is 64-bit. The seed in this function is very different from all other MurmurHash variants.

There is a sense that any seed under `0x100000000` is not very good, because it only affects h1 and not h2. So ideally, you would want data in both uppper 32-bit and lower 32-bit of seed.

Since I ported this to JS, I cannot do this: `uint32_t(seed >> 32);` The solution I came up with is to split `uint64_t seed` into two 32-bit integers, `seed1` and `seed2`, acting in same way. 

It is interesting that `MurmurHash64A` and `MurmurHash64B` are the only functions that take 64-bit seed argument. And MurmurHash64B is specifically meant for 32-bit platforms. 

Also, are you aware of the flaw in MurmurHash64B ? It is not a true 64-bit hash since if input stripe of `h1` is damaged, it does not affect the output of `h2`. In these situations the collision probability is same as 32-bit.  I think this can be fixed by XORing `h1` and `h2` while reading input.
